examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2755 @@
+  // returns a pair of values, one each for the $s$ and $z$ slack
+  // variables.
+  template <int dim>
+  std::pair<double, double> SANDTopOpt<dim>::calculate_max_step_size(
+    const BlockVector<double> &state,
+    const BlockVector<double> &step) const
+  {
+    double       fraction_to_boundary;
+    const double min_fraction_to_boundary = .8;
+    const double max_fraction_to_boundary = .99999;
+
+    if (min_fraction_to_boundary < 1 - barrier_size)
+      {
+        if (1 - barrier_size < max_fraction_to_boundary)
+          fraction_to_boundary = 1 - barrier_size;
+        else
+          fraction_to_boundary = max_fraction_to_boundary;
+      }
+    else
+      fraction_to_boundary = min_fraction_to_boundary;
+
+    double step_size_s_low  = 0;
+    double step_size_z_low  = 0;
+    double step_size_s_high = 1;
+    double step_size_z_high = 1;
+    double step_size_s, step_size_z;
+
+    for (unsigned int k = 0; k < 50; ++k)
+      {
+        step_size_s = (step_size_s_low + step_size_s_high) / 2;
+        step_size_z = (step_size_z_low + step_size_z_high) / 2;
+
+        const BlockVector<double> state_test_s =
+          (fraction_to_boundary * state) + (step_size_s * step);
+
+        const BlockVector<double> state_test_z =
+          (fraction_to_boundary * state) + (step_size_z * step);
+
+        const bool accept_s =
+          (state_test_s.block(SolutionBlocks::density_lower_slack)
+             .is_non_negative()) &&
+          (state_test_s.block(SolutionBlocks::density_upper_slack)
+             .is_non_negative());
+        const bool accept_z =
+          (state_test_z.block(SolutionBlocks::density_lower_slack_multiplier)
+             .is_non_negative()) &&
+          (state_test_z.block(SolutionBlocks::density_upper_slack_multiplier)
+             .is_non_negative());
+
+        if (accept_s)
+          step_size_s_low = step_size_s;
+        else
+          step_size_s_high = step_size_s;
+
+        if (accept_z)
+          step_size_z_low = step_size_z;
+        else
+          step_size_z_high = step_size_z;
+      }
+
+    return {step_size_s_low, step_size_z_low};
+  }
+
+
+  // @sect4{Computing residuals}
+
+  // The next function computes a right hand side vector linearized
+  // around a "test solution vector" that we can use to look at the
+  // magnitude of the KKT conditions.  This is then used for testing
+  // the convergence before shrinking the barrier size, as well as in the
+  // calculation of the $l_1$ merit.
+  //
+  // The function is lengthy and complicated, but it is really just a
+  // copy of the right hand side part of what the `assemble_system()`
+  // function above did.
+  template <int dim>
+  BlockVector<double> SANDTopOpt<dim>::calculate_test_rhs(
+    const BlockVector<double> &test_solution) const
+  {
+    BlockVector<double> test_rhs;
+    test_rhs.reinit(
+      system_rhs); /* a zero vector with size and blocking of system_rhs */
+
+    const FEValuesExtractors::Scalar densities(
```

Comment:
you are using these extractors more than once. Maybe move them into a namespace?